### DISCUSSION TITLE
Create C bindings

### DISCUSSION THREAD

**Discussion by Johannesd3** [librespot-org/librespot#689] [Feature Requests]
@d4rwel inspired me to write down some thoughts I had about creating C bindings for librespot. There's no need to change anything in librespot itself, it could happen in its own crate/repo. I assume that #665 is ready, but it could also be accomplished with git dependencies.

So here's a rough plan:

Create a rust library crate and add the following to the `Cargo.toml`:

```toml
[dependencies]
librespot = "<choose version>"
tokio = "<choose version>"

[lib]
crate-type = ["cdylib"]
``` 

Pick some important Rust functions from librespot. Maybe look at the examples to decide what's needed.

Create functions that are callable from C as shown in ~[this gist](https://gist.github.com/Johannesd3/2963d251945347de8988585d61e1e9b5)~, or rather [this repository](https://github.com/Johannesd3/c-bindings-example).

Important: Wrap everything into a [catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html). If an error occurs, there might be UB if it reaches the border to C.

There are certainly some structs that need to be passed. Since those structs are not compatible with C, they must be passed as opaque type behind a pointer (as in the example).

Some functions are `async`. These functions must be executed by a [tokio runtime](https://docs.rs/tokio/1.4.0/tokio/runtime/struct.Runtime.html#method.block_on). Create functions to created and destroy a tokio runtime. Return this runtime as opaque struct. (This struct could even be called `librespot` if the user shouldn't be bothered with implementation details.) For every function call that executes an async rust function, a pointer to this struct must be passed.

Remember, destructors in Rust are implicit. It's important to create functions that deallocate resources again, although these functions seem to do nothing. Look at the example above.

If it's ready, running `cargo build` will produce a `.so` file in `target/debug`.

Run `cargo install cbindgen` to install a tool to create a C header file. `cbindgen --lang c` will hopefully spit out the expected result.

You can now write a C application that uses librespot. Just include the header and link it against this `.so` file.

How does it sound?

**Comment by sashahilton00** [comment]
I believe C bindings were discussed at some point a few years ago, but work never really progressed on them. My thoughts on it are limited since I have no need of them/plan to use them, though can understand why they might be useful. I think any development should happen in a completely seperate repository though, purely for separation of concerns.

**Reply by kingosticks** [reply]
I'm not sure I agree these are separate concerns. Bindings to a library are fundamentally entwined with the lib's API and changes should be considered.

**Reply by Johannesd3** [reply]
They're not really entwined. The C bindings would be just an additional layer on top of the Rust API. Development of Librespot should happen independently. If someone needs C bindings to the latest librespot version, the dependency to librespot could just be bumped in its own repository.

**Reply by TheNoim** [reply]
Well, why do you make it so hard to use librespot? Why not just provide c bindings? This would be the most universal way to make this library available for all languages. This decision just makes it hard for not so experienced programmers with rust.

**Reply by Johannesd3** [reply]
> just

There's some work necessary, and we didn't find someone willing to do it.

**Reply by Thalhammer** [reply]
Hi I am new to this repo, but with the shutdown of libspotify's server side on 16th May I am looking for an alternative and this seems to be one of the best maintained ones. However none of my code uses rust right now (nor will it ever) so my options are either:
Somehow write a binding layer like proposed here, or rewriting the entire library (or what I need) in C++.
This raises a question:
If I where to rewrite it in C++, would there be any hope in getting it under the "librespot-org" roof (similar to librespot-java) ?

**Reply by roderickvd** [reply]
To better understand your question, why do you feel it's important to be adopted by `librespot-org`?

**Reply by Thalhammer** [reply]
It isn't hugely important for me, but it feels like the right thing to do given it would probably be more or less a direct reimplementation of the existing libs, just in a new language.

**Reply by roderickvd** [reply]
I don't know about a policy nor is it up to me. But generally I'd say that all well-maintained ports would make good candidates.

**Reply by sashahilton00** [reply]
We can create a namespace and a team for the C++ version if that would be of interest. Only thing I would say is that it might be better to implement a basic version in your own repo, then assuming it gets to the point where it has enough functionality/stability to be considered a viable client/library, and you're happy to be the ongoing maintainer of that version, it can be transferred/renamed to `librespot-cpp` under the `librespot-org` organisation.

**Reply by Thalhammer** [reply]
> Only thing I would say is that it might be better to implement a basic version in your own repo, then assuming it gets to the point where it has enough functionality/stability to be considered a viable client/library, and you're happy to be the ongoing maintainer of that version, it can be transferred/renamed to `librespot-cpp` under the `librespot-org` organisation.

Thats what I had in mind. I just asked beforehand because I noticed over time that some organisations seem to be quite reluctant against transfering in external projects, especially if its not in one of the "cool" and modern languages.
